### ISSUE TITLE
Redeclaration across multiple scripts between lexical and Annex B function throws SyntaxError on all implementation, but not on spec

### ISSUE THREAD

**Issue by arai-a** [tc39/ecma262#2019]
The following HTML codelet throws SyntaxError, on all engines:

```
<script>
let f;
</script>
<script>
{
  function f() {}
}
</script>
```

but I cannot locate any spec step that does so.

https://tc39.es/ecma262/#sec-web-compat-globaldeclarationinstantiation

Annex B.3.3.2 step 2.d.ii.1 says "If env.HasLexicalDeclaration(F) is false, then".

Given that the `env` has lexical declaration `f` created by the first script (am I correct?),
the remaining steps won't apply to the function `f` in the second script.
so the binding for function `f` will be created in the lexical scope for the block, and won't conflict with global lexical binding.

And even if the it applies somehow, the steps that detects the redeclaration between scripts are 15.1.11 steps 4-5 (again, am I correct?), that's before Annex B 3.3.2, that's performed at step 13.

https://tc39.es/ecma262/#sec-globaldeclarationinstantiation
```
4. For each name in lexNames, do
  a. If env.HasVarDeclaration(name) is true, throw a SyntaxError exception.
  b. If env.HasLexicalDeclaration(name) is true, throw a SyntaxError
     exception.
  c. Let hasRestrictedGlobal be ? env.HasRestrictedGlobalProperty(name).
  d. If hasRestrictedGlobal is true, throw a SyntaxError exception.
5. For each name in varNames, do
  a. If env.HasLexicalDeclaration(name) is true, throw a SyntaxError
     exception.
...
13. NOTE: Annex B.3.3.2 adds additional steps at this point.
```

**Comment by bakkot** [comment]
I agree with your analysis. I believe the intent is for that snippet to be legal, so this would be a bug in all engines. (Unless there is something in the HTML spec which would cause this to be a syntax error, but I do not believe there is.)

It is also worth noting that

```html
<script>
let f;
</script>
<script>
'use strict';
{
  function f() {}
}
</script>
```
does *not* throw.

cc @syg

**Comment by syg** [comment]
I did some more thinking and it seems to me that this is actually difficult to fix in engines.

An implementation really wants to make the Annex B.3.3 decision at parse time, because it needs to emit the extra assignment to the synthetic var, and actually declare the var binding at parse time. At the same time, off-thread (and possibly parallel) parsing due to optimizations around `async` and `defer` scripts means that you _don't have_ an up-to-date picture of the global bindings until all parsing threads are joined and their top-level binding tables reconciled.

Fixing this might take a while, without a lot to gain for it. It is unlikely to be prioritized anytime soon.

**Comment by bathos** [comment]
> All the engines here are wrong
> Fixing this might take a while, without a lot to gain for it

could the solution be the engines _becoming_ right?
